include/extractor/restriction_index.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef OSRM_EXTRACTOR_RESTRICTION_INDEX_HPP_
+#define OSRM_EXTRACTOR_RESTRICTION_INDEX_HPP_
+
+#include "extractor/restriction.hpp"
+#include "util/typedefs.hpp"
+
+#include <boost/unordered_map.hpp>
+
+#include <utility>
+#include <vector>
+
+namespace osrm
+{
+namespace extractor
+{
+
+// allows easy check for whether a node intersection is present at a given intersection
+template <typename restriction_type> class RestrictionIndex
+{
+  public:
+    using value_type = restriction_type;
+
+    template <typename extractor_type>
+    RestrictionIndex(std::vector<restriction_type> &restrictions, extractor_type extractor);
+
+    bool IsIndexed(NodeID first, NodeID second) const;
+
+    auto Restrictions(NodeID first, NodeID second) const
+    {
+        return restriction_hash.equal_range(std::make_pair(first, second));
+    };
+
+    auto Size() const { return restriction_hash.size(); }
+
+  private:
+    boost::unordered_multimap<std::pair<NodeID, NodeID>, restriction_type *> restriction_hash;
```

Comment:
Is there any preference for `boost::unordered_multimap` over `std::unordered_multimap`